File: leetcode/python/0350_intersection_of_two_arrays_II.py
class Solution(object):
    def intersect(self, nums1, nums2):
        """
        :type nums1: List[int]
        :type nums2: List[int]
        :rtype: List[int]
        """
        nums1Count = {}
        for num1 in nums1:
            if not num1 in nums1Count: nums1Count[num1] = 1
            else: nums1Count[num1] += 1

        res = []
        for num2 in nums2:
            if num2 in nums1Count:
                if nums1Count[num2] == 1: nums1Count.pop(num2)
                else: nums1Count[num2] -= 1
                res.append(num2)
        
        return res

File: leetcode/cpp/0350_intersection_of_two_arrays_II.cpp
class Solution {
public:
    vector<int> intersect(vector<int>& nums1, vector<int>& nums2) {
        if(nums1.size() > nums2.size()){
            swap(nums1, nums2);
        }
        
        unordered_map<int, int> m;
        for(auto val : nums1){
            m[val]++;
        }
        
        int k = 0;
        
        for(auto val : nums2){
            if(m[val] > 0){
                nums1[k++] = val;
                --m[val];
            }
        }
        
        return vector<int>(nums1.begin(), nums1.begin() + k);
    }
};